src/components/Launcher.jsx

```diff
@@ -0,0 +1,188 @@
+import React, { Component } from 'react'
+import { translate } from '../lib/I18n'
+import Autosuggest from 'react-autosuggest';
+import { rawIntent } from '../lib/intents'
+
+const INTENT_VERB = 'OPEN'
+const INTENT_DOCTYPE = 'io.cozy.suggestions'
+
+class Launcher extends Component {
+  constructor (props) {
+    super(props)
+    
+    this.state = {
+      query: '',
+      suggestions: []
+    }
+    
+    this.sourcesNode = null
+    this.sources = []
+  }
+  
+  componentDidMount () {
+    // The launcher has one or more sources that provide suggestions. These sources are iframes into other apps, provied by thee intent system.
+    // Since we need to call the sources whenever the query changes, we are taking manual control over the intent process.
+    window.addEventListener('message', this.onMessageFromSource)
+    
+    rawIntent(INTENT_VERB, INTENT_DOCTYPE)
+    .then(intent => {
+      const { services } = intent.attributes
+      if (!services) return
+      
+      services.forEach(service => {
+        // we're injecting the iframes, but they will stay invisible
+        const iframe = document.createElement('iframe')
+        const url = service.href
+        iframe.setAttribute('src', url)
+        iframe.style.display = 'none'
+        this.sourcesNode.appendChild(iframe)
+        const serviceOrigin = url.split('/', 3).join('/')
+        
+        this.sources.push({
+          slug: service.slug,     // can be used to show where a suggestion comes from
+          origin: serviceOrigin,
+          id: intent._id,
+          ready: false,
+          window: null,           // will hold a reference to the window we're sending messages to
+          resolve: null           // a reference to a function to call when the source sends suggestions
+        })
+      })
+    })
+  }
+  
+  onMessageFromSource = (event) => {
+    // this re-implements a subset of injectService found in lib/intents, though only the part that are useful for suggestions
+    const source = this.sources.find(source => source.origin === event.origin)
+    
+    if (!source) return
+    
+    if (event.data.type === `intent-${source.id}:ready`) {
+      source.ready = true
+      source.window = event.source
+      
+      source.window.postMessage({}, event.origin)
+    }
+    else if (event.data.type === `intent-${source.id}:data` && source.resolve) {
+      source.resolve({
+        id: source.id,
+        suggestions: event.data.suggestions
+      })
+      source.resolve = null
+    }
+    else {
+      console.log('unhandled message:', event)
+    }
+  }
+  
+  onChange = (event, { newValue }) => {
+    this.setState({
+      query: newValue
+    })
```

Comment:
Could we take the habit to use the functional form of `setState`:

```js
this.setState((state) => ({...state, query: newValue})
```

So `onChange` could be written:

```js
onChange = (event, { newValue: query }) => {
  this.setState(state => ({ ...state, query })
}
```

Reply:
Is there an upside to using the functional form when the new state is *not* computed from the previous state?

Reply:
The very important upside is `setState` is asynchronous and multiple sequential calls could mislead to unwanted state.
And the corollary is it uses pure function which are easier to maintain.

Even if you don't mutate the state, even if you are quite sure you are not calling `setState` multiple times, you may be homogeneous with the rest of your code and to always use the same syntax/format.

That is why I recommend using the functional `setState` instead of the shorthand version.

Reply:
I understand the upside, but in this case the upside doesn't apply? Sequential calls would be perfectly fine here, we always want the latest version of `query` as provided by the AutoSuggest component.

And in terms of homogeneity, I think our codebase favors the shorthand version. I also find it more readable.